driver/src/main/scala/com/datastax/spark/connector/cql/Schema.scala

```diff
@@ -398,32 +618,28 @@ object Schema extends StrictLogging {
       }
 
     def fetchKeyspaces(metadata: Metadata): Set[KeyspaceDef] =
-      for ((_, keyspace) <- metadata.getKeyspaces.asScala.toSet if isKeyspaceSelected(keyspace)) yield
-        fetchKeyspace(keyspace, tableName)
+      metadata.getKeyspaces.values().asScala
+        .filter(isKeyspaceSelected(_))
+        .map(KeyspaceDef(_, tableName))
+        .toSet
 
     logger.debug(s"Retrieving database schema")
-    def fetchSchema(metadata: => Metadata): Schema =
-      Schema(fetchKeyspaces(metadata))
-
-    val scheme = fetchSchema(session.refreshSchema())
-
-    logger.debug(s"${scheme.keyspaces.size} keyspaces fetched: " +
-      s"${scheme.keyspaces.map(_.keyspaceName).mkString("{", ",", "}")}")
-    scheme
+    val schema = Schema(fetchKeyspaces(session.refreshSchema()))
+    logger.debug(s"${schema.keyspaces.size} keyspaces fetched: " +
+      s"${schema.keyspaces.map(_.keyspaceName).mkString("{", ",", "}")}")
+    schema
   }
 
-
   /**
-    * Fetches a TableDef for a particular Cassandra Table throws an
-    * exception with name options if the table is not found.
-    */
-  def tableFromCassandra(
-      session: CqlSession,
-      keyspaceName: String,
-      tableName: String): TableDef = {
+   * Fetches a TableDef for a particular Cassandra Table throws an
+   * exception with name options if the table is not found.
+   */
+  def tableFromCassandra(session: CqlSession,
+                         keyspaceName: String,
+                         tableName: String): DriverTableDef = {
 
     fromCassandra(session, Some(keyspaceName), Some(tableName)).tables.headOption match {
```

Comment:
I agree that the use of headOption here is a bit cumbersome, but unfortunately we're stuck with fromCassandra() for the foreseeable future.  There are a fair amount of users who access that functionality through com.datastax.spark.connector.schemaFromCassandra(), which is used in a number of places throughout the code base.  It's probably cleaner to re-work this API to return KeyspaceDefs or TableDefs directly rather than using Schema as a common return type (which in turn requires this "Schema with only some keyspaces/tables" nonsense used here).  That seemed like a larger and/or discrete effort, however... and this ticket is already big enough. :)